File: q0880/q0880.java
class Solution {
    public String decodeAtIndex(String s, int k) {
        long totalLen = 0;

        // go forward
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isDigit(c))
                totalLen *= Integer.parseInt(String.valueOf(c));
            else
                totalLen++;
        }

        // reverse to look for the character
        for (int i = s.length() - 1; i >= 0; i--) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                totalLen /= Integer.parseInt(String.valueOf(c));
                k %= totalLen;
            } else {
                if (k == 0 || totalLen == k)
                    return String.valueOf(c);
                totalLen--;
            }
        }
        return "";
    }
}

File: q0880/q0880.cpp
class Solution
{
public:
    string decodeAtIndex(string s, int k)
    {
        long long totalLen = 0;
        for (char c : s)
        {
            if (isdigit(c))
                totalLen *= c - '0';
            else
                totalLen++;
        }

        for (int i = s.size() - 1; i >= 0; i--)
        {
            char c = s[i];
            if (isdigit(c))
            {
                totalLen /= (c - '0');
                k = k % totalLen;
            }
            else
            {
                if (k == 0 || totalLen == k)
                    return string("") + c;
                totalLen--;
            }
        }
        return "";
    }
};